File: leetcode/C++/search-a-2d-matrix.cpp
class Solution {
public:
    int rowSearch(vector<vector<int>>& matrix, int &target){
        int l=0, r=(int)matrix.size()-1,m=0, sol=0;
        while(l<=r){
            m=(l+r)/2;
            if(matrix[m][0]>target) r=m-1;
            else {
                sol=m;
                l=m+1;
            }
        }
        return sol;
    }
    bool colSearch(vector<int> &row, int& target){
        int l=0, r=(int)row.size()-1, m=0;
        while(l<=r){
            m=(l+r)/2;
            if(row[m]>target) r=m-1;
            else if(row[m]<target) l=m+1;
            else return true;
        }
        return false;
    }
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int r = rowSearch(matrix, target);
        return colSearch(matrix[r], target);
    }
};

File: leetcode/Python/search-a-2d-matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        def rowSearch():
            l, r, m = 0, len(matrix)-1, 0
            sol = l
            while l<=r:
                m = (l+r)//2
                if matrix[m][0]>target:
                    r = m - 1
                else:
                    sol, l = m, m+1
            return sol
        def colSearch(row):
            l, r, m = 0, len(row)-1, 0
            while l<=r:
                m = (l+r)//2
                if row[m] > target:
                    r = m - 1
                elif row[m] < target:
                    l = m + 1
                else:
                    return True
            return False
        return colSearch(matrix[rowSearch()])